File: APPLICATION/models/detalle_venta.py
"""
detalle_venta.py — Modelo de Detalle de Venta con serialización binaria.

Estructura del registro (33 bytes por registro):
┌──────────┬──────────────┬──────────────┬──────────────┬──────────────────┬──────────────┬──────────┐
│  id      │  venta_id    │ producto_id  │  cantidad    │ precio_unitario  │  subtotal    │  estado  │
│  4 bytes │  4 bytes     │  4 bytes     │  4 bytes     │  8 bytes         │  8 bytes     │  1 byte  │
└──────────┴──────────────┴──────────────┴──────────────┴──────────────────┴──────────────┴──────────┘
Total: 4 + 4 + 4 + 4 + 8 + 8 + 1 = 33 bytes
"""
import struct
from dataclasses import dataclass
from typing import ClassVar


@dataclass
class DetalleVenta:
    """
    Representa una línea de detalle de una venta (producto + cantidad).

    Serialización: struct big-endian "!IIIIddB"
    Tamaño fijo:   33 bytes por registro
    """

    # ! = big-endian
    # I = unsigned int (4 bytes) × 4 → id, venta_id, producto_id, cantidad
    # d = double       (8 bytes) × 2 → precio_unitario, subtotal
    # B = unsigned char (1 byte)     → estado
    FORMAT: ClassVar[str] = "!IIIIddB"
    SIZE:   ClassVar[int] = struct.calcsize("!IIIIddB")  # 33

    id:              int   = 0
    venta_id:        int   = 0
    producto_id:     int   = 0
    cantidad:        int   = 0
    precio_unitario: float = 0.0
    subtotal:        float = 0.0
    estado:          int   = 1

    def pack(self) -> bytes:
        """Serializa el detalle a 33 bytes de longitud fija."""
        return struct.pack(
            self.FORMAT,
            self.id,
            self.venta_id,
            self.producto_id,
            self.cantidad,
            self.precio_unitario,
            self.subtotal,
            self.estado,
        )

    @classmethod
    def unpack(cls, data: bytes) -> "DetalleVenta":
        """Deserializa 33 bytes a un objeto DetalleVenta."""
        if len(data) != cls.SIZE:
            raise ValueError(
                f"DetalleVenta.unpack: se esperaban {cls.SIZE} bytes, "
                f"se recibieron {len(data)}"
            )
        raw = struct.unpack(cls.FORMAT, data)
        return cls(
            id=raw[0], venta_id=raw[1], producto_id=raw[2], cantidad=raw[3],
            precio_unitario=raw[4], subtotal=raw[5], estado=raw[6],
        )

    def to_dict(self) -> dict:
        return {
            "id": self.id, "venta_id": self.venta_id,
            "producto_id": self.producto_id, "cantidad": self.cantidad,
            "precio_unitario": self.precio_unitario,
            "subtotal": self.subtotal, "estado": self.estado,
        }

    def to_csv_row(self) -> str:
        return (
            f"{self.id},{self.venta_id},{self.producto_id},"
            f"{self.cantidad},{self.precio_unitario:.2f},{self.subtotal:.2f},{self.estado}"
        )

    def __repr__(self) -> str:
        return (
            f"DetalleVenta(id={self.id}, venta={self.venta_id}, "
            f"prod={self.producto_id}, cant={self.cantidad}, "
            f"sub=S/{self.subtotal:.2f})"
        )

File: APPLICATION-JAVA/src/main/java/com/minimarket/models/DetalleVenta.java
package com.minimarket.models;

import java.io.*;

/**
 * Modelo de DetalleVenta con serialización binaria de longitud fija.
 *
 * Estructura del registro (33 bytes):
 *   id(4B) + venta_id(4B) + producto_id(4B) + cantidad(4B)
 *   + precio_unitario(8B) + subtotal(8B) + estado(1B)
 *
 * Compatible con Python: struct.pack("!IIIIddB", ...)
 */
public class DetalleVenta implements BinaryRecord {

    public static final int SIZE = 33; // 4 + 4 + 4 + 4 + 8 + 8 + 1

    private int    id;
    private int    ventaId;
    private int    productoId;
    private int    cantidad;
    private double precioUnitario;
    private double subtotal;
    private int    estado;

    public DetalleVenta() { this.estado = 1; }

    public DetalleVenta(int id, int ventaId, int productoId, int cantidad,
                        double precioUnitario, double subtotal, int estado) {
        this.id             = id;
        this.ventaId        = ventaId;
        this.productoId     = productoId;
        this.cantidad       = cantidad;
        this.precioUnitario = precioUnitario;
        this.subtotal       = subtotal;
        this.estado         = estado;
    }

    // ── Serialización ────────────────────────────────────────────────────────

    @Override
    public byte[] toBytes() {
        ByteArrayOutputStream bos = new ByteArrayOutputStream(SIZE);
        DataOutputStream dos = new DataOutputStream(bos);
        try {
            dos.writeInt(id);
            dos.writeInt(ventaId);
            dos.writeInt(productoId);
            dos.writeInt(cantidad);
            dos.writeDouble(precioUnitario);
            dos.writeDouble(subtotal);
            dos.writeByte(estado);
        } catch (IOException e) {
            throw new RuntimeException("Error serializando DetalleVenta", e);
        }
        return bos.toByteArray();
    }

    public static DetalleVenta fromBytes(byte[] data) throws IOException {
        if (data.length != SIZE) {
            throw new IllegalArgumentException(
                "DetalleVenta.fromBytes: se esperaban " + SIZE + " bytes, se recibieron " + data.length);
        }
        DataInputStream dis = new DataInputStream(new ByteArrayInputStream(data));
        DetalleVenta d = new DetalleVenta();
        d.id             = dis.readInt();
        d.ventaId        = dis.readInt();
        d.productoId     = dis.readInt();
        d.cantidad       = dis.readInt();
        d.precioUnitario = dis.readDouble();
        d.subtotal       = dis.readDouble();
        d.estado         = dis.readByte() & 0xFF;
        return d;
    }

    // ── Exportación ──────────────────────────────────────────────────────────

    public String toCsvRow() {
        return String.format("%d,%d,%d,%d,%.2f,%.2f,%d",
            id, ventaId, productoId, cantidad, precioUnitario, subtotal, estado);
    }

    // ── BinaryRecord ─────────────────────────────────────────────────────────

    @Override public int getId()           { return id; }
    @Override public void setId(int id)    { this.id = id; }
    @Override public int getEstado()       { return estado; }
    @Override public void setEstado(int e) { this.estado = e; }

    // ── Getters / Setters ────────────────────────────────────────────────────

    public int    getVentaId()              { return ventaId; }
    public void   setVentaId(int v)         { ventaId = v; }
    public int    getProductoId()           { return productoId; }
    public void   setProductoId(int v)      { productoId = v; }
    public int    getCantidad()             { return cantidad; }
    public void   setCantidad(int v)        { cantidad = v; }
    public double getPrecioUnitario()       { return precioUnitario; }
    public void   setPrecioUnitario(double v) { precioUnitario = v; }
    public double getSubtotal()             { return subtotal; }
    public void   setSubtotal(double v)     { subtotal = v; }

    @Override
    public String toString() {
        return String.format("DetalleVenta(id=%d, venta=%d, prod=%d, cant=%d, sub=S/%.2f)",
            id, ventaId, productoId, cantidad, subtotal);
    }
}
